File: JexFlix-Scraper/Flixify/Flixify.cs
﻿using BrotliSharpLib;
using Newtonsoft.Json;
using System;
using System.Collections.Specialized;
using System.IO;
using System.Net;
using System.Text;
using System.Threading;

namespace JexFlix_Scraper.Flixify {

    public static class Flixify {

        private static CookieAwareWebClient Web = new CookieAwareWebClient();
        private const string FLIXIFY = "https://flixify.com/";
        private const string MOVIES_URL = FLIXIFY + "movies?_t=limjml&_u=ji9joxc5ip&add_mroot=1&description=1&g={0}&o=t&p={1}&postersize=poster&previewsizes=%7B%22preview_list%22:%22big3-index%22,%22preview_grid%22:%22video-block%22%7D&slug=1&type=movies";
        private const string MOVIES_URL_FOR_DOWNLOAD = FLIXIFY + "{0}?_t=lmispq&_u=ji9joxc5ip&add_mroot=1&cast=0&crew=0&description=1&episodes_list=1&has_sequel=1&postersize=poster&previews=1&previewsizes=%7B%22preview_grid%22:%22video-block%22,%22preview_list%22:%22big3-index%22%7D&season_list=1&slug=1&sub=1";

        public static void Run() {

            // bypass cloudflare so we can login to and access the website
            Networking.BypassCloudFlare(FLIXIFY + "/login");

            // initialize request headers
            Web.Headers.Add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36");
            Web.Headers.Add("Accept-Encoding", "gzip, deflate, br");
            Web.Headers.Add("Accept-Language", "en-US,en;q=0.9,ja;q=0.8");

            // establish post data
            NameValueCollection values = new NameValueCollection();
            values["ref"] = "";
            values["email"] = "nex@weebware.net";
            values["password"] = "fuckniggers69";

            // send request to store cookies from valid login
            Web.UploadValues(FLIXIFY + "/login", values);

            InitializeScraper();

        }

        public static void InitializeScraper() {

            foreach (string genre in genres) {

                for (int page = 1; page <= 100; page++) {

                    // apply headers to web request
                    Web.FlixifyHeaders();
                    byte[] response = null;

                    try {
                        string url = string.Format(MOVIES_URL, genre, page);
                        response = Web.DownloadData(url);
                    } catch (WebException ex) {
                        // catch NotFound exception
                        // continue to next genre (out of videos)
                        HttpWebResponse webResponse = ex.Response as HttpWebResponse;
                        if (webResponse.StatusCode == HttpStatusCode.NotFound)
                            break;
                    }

                    // decompress and establish response from server
                    byte[] decompressed = Brotli.DecompressBuffer(response, 0, response.Length);
                    string raw = Encoding.Default.GetString(decompressed);

                    // parse video list
                    ParseMovies(raw);

                    Thread.Sleep(1000);

                }

            }

        }

        public const string UPLOAD_URL = "https://cdn.jexflix.com";
        public static void ParseMovies(string raw) {

            MovieData serverData = JsonConvert.DeserializeObject<MovieData>(raw);

            foreach (Movie movie in serverData.items) {

                // if it already exists on the server, go to next movie
                if (Networking.FileExists(movie.title))
                    continue;

                Web.FlixifyHeaders();

                byte[] response = Web.DownloadData(string.Format(MOVIES_URL_FOR_DOWNLOAD, movie.url));
                byte[] response_decompressed = Brotli.DecompressBuffer(response, 0, response.Length);
                string new_raw = Encoding.Default.GetString(response_decompressed);

                RootObject rootObject = JsonConvert.DeserializeObject<RootObject>(new_raw);

                Data data = new Data();
                data.title = rootObject.item.title;
                data.url = rootObject.item.url.Substring(8);
                data.description = rootObject.item.description;
                data.duration = rootObject.item.duration;
                data.thumbnail = UPLOAD_URL + rootObject.item.url + "/thumbnail.jpg";
                data.preview = UPLOAD_URL + rootObject.item.url + "/preview.jpg";
                data.genres = rootObject.item.genres;
                data.imdb_id = rootObject.item.imdb_id;
                data.year = rootObject.item.year;
                data.certification = rootObject.item.certification;

                // setup qualities
                if (rootObject.item.download.download_720 != null) data.qualities.Add(new Qualities { resolution = 720, link = UPLOAD_URL + rootObject.item.url + "/720.mp4" });
                if (rootObject.item.download.download_1080 != null) data.qualities.Add(new Qualities { resolution = 1080, link = UPLOAD_URL + rootObject.item.url + "/1080.mp4" });

                // upload info to insert into database
                Web.UploadString("https://scraper.jexflix.com/add_movie.php", JsonConvert.SerializeObject(data));

                DownloadFiles(rootObject);
                //Networking.UploadFiles(rootObject);
                Directory.Delete(rootObject.item.title.Sanitized(), true);

                Console.WriteLine("Successfully uploaded all data for: " + data.title + Environment.NewLine);

            }

        }

        public const string BASE_IMAGES_URL = "https://a.flixify.com";
        public const string BASE_URL = "https://flixify.com";
        public static void DownloadFiles(RootObject data) {
            CookieAwareWebClient web = new CookieAwareWebClient();
            // create directory to download the files to, we will delete this later.
            string directory = data.item.title.Sanitized();
            if (!Directory.Exists(directory)) Directory.CreateDirectory(directory);

            string preview_url = BASE_IMAGES_URL + data.item.images.preview_large;
            string thumbnail_url = BASE_IMAGES_URL + data.item.images.poster;

            Console.WriteLine("Downloading " + data.item.title + " preview...");
            if (!File.Exists(directory + "/preview.jpg")) web.DownloadFile(preview_url, directory + "/preview.jpg");
            Console.WriteLine("Completed.");
            Console.WriteLine("Downloading " + data.item.title + " thumbnail...");
            if (!File.Exists(directory + "/thumbnail.jpg")) web.DownloadFile(thumbnail_url, directory + "/thumbnail.jpg");
            Console.WriteLine("Completed.");

            if (data.item.download.download_720 != null) {
                Console.WriteLine("Downloading " + data.item.title + " in 720p...");
                if (!File.Exists(directory + "/720.mp4")) web.DownloadFile(BASE_URL + data.item.download.download_720, directory + "/720.mp4");
                Console.WriteLine("Completed.");
            }
            if (data.item.download.download_1080 != null) {
                Console.WriteLine("Downloading " + data.item.title + " in 1080p...");
                if (!File.Exists(directory + "/1080.mp4")) web.DownloadFile(BASE_URL + data.item.download.download_1080, directory + "/1080.mp4");
                Console.WriteLine("Completed.");
            }
        }

        public static void FlixifyHeaders(this CookieAwareWebClient web) {
            web.Headers.Clear(); // clear any existing headers
            web.Headers.Add("Host", "flixify.com");
            web.Headers.Add("Accept", "application/json");
            web.Headers.Add("User-Agent", Networking.USER_AGENT);
            web.Headers.Add("Referer", "https://flixify.com/movies?_rsrc=chrome/newtab");
        }

        private static string[] genres = {
            "animation",
            "documentary",
            "fantasy",
            "music",
            "science-fiction",
            "western",
            "action",
            "comedy",
            "drama",
            "history",
            "mystery",
            "thriller",
            "adventure",
            "crime",
            "family",
            "horror",
            "romance",
            "war"
            };

    }

}

File: scraper/scraper/flixify.py
# import selenium, for original login
from selenium import webdriver

# import general modules
import urllib, time, json, requests, threading
import database, utils


# flixify urls
SITE_URL = "https://calmx.site/"
ASSETS_URL = "https://a.calmx.site/"

# whether or not we should hide the browser window when logging in
HIDE_BROWSER = True

# list of genres on flixify
GENRES = (
    "animation",            # 0
    "fantasy",              # 1
    "science-fiction",      # 2
    "music",                # 3
    "documentary",          # 4
    "western",              # 5
    "action",               # 6
    "comedy",               # 7
    "drama",                # 8
    "history",              # 9
    "mystery",              # 10
    "thriller",             # 11
    "adventure",            # 12
    "crime",                # 13
    "family",               # 14
    "horror",               # 15
    "romance",              # 16
    "war"                   # 17
    )

class flixify:

    def __init__(self, email, password):

        # initializes webdriver
        # note: firefox uses geckodriver, so it must be in PATH
        # https://github.com/mozilla/geckodriver
        print("Initializing FireFox driver...", end = " ")
        options = webdriver.firefox.options.Options()
        options.set_headless(HIDE_BROWSER)
        browser = webdriver.Firefox(firefox_options = options)
        print("done")

        # store user-agent for future use
        self.user_agent = browser.execute_script("return navigator.userAgent")

        # load login page
        print("Loading Flixify...", end = " ")
        browser.get(SITE_URL + "login")
        print("done")

        # login to the website (set username and password, click submit)
        print("Logging in...")
        browser.find_elements_by_name("email")[0].send_keys(email)
        browser.find_elements_by_name("password")[0].send_keys(password)
        browser.find_element_by_xpath("//input[@value='LOGIN']").click()

        # determine '_t' and '_u' variables
        # these are loaded via javascript, so we may have to wait
        while True:

            # parse url and determine query parameter
            url_info = urllib.parse.urlparse(browser.current_url)
            query_info = urllib.parse.parse_qs(url_info.query)

            # if we're missing '_t' or '_u', wait a second and try again
            if not '_t' in query_info or not '_u' in query_info:
                time.sleep(1)
                continue

            # store values and break out of loop
            self.var_t = query_info['_t'][0]
            self.var_u = query_info['_u'][0]
            break

        self.referer_url = browser.current_url
        print("Logged in successfully!")
        print("_t variable: " + self.var_t)
        print("_u variable: " + self.var_u)

        self.session = utils.session_from_driver(browser)
        browser.quit()

        self.headers = {
            "User-Agent": self.user_agent,
            "Referer": self.referer_url,
            "Accept": "application/json",
            "TE": "Trailers"
        }

    def build_list_url(self, type, page, genre = None):
        """
        Builds a flixify api url from a type & page.
        This will retrieve a list of 'items', aka movies/shows.

        Parameters:
            type (string): Type of content. ('movies' or 'shows')
            page (int): Page to get items from.
            genre (string): Genre to get videos from.

        Returns:
            string: The url of the api to send a request to.
        """

        # build standard parameters
        params = {
            "_t": self.var_t,
            "_u": self.var_u,
            "add_mroot": 1,
            "description": 1,
            "o": "t",
            "p": page,
            "postersize": "poster",
            "previewsizes": '{"preview_list":"big3-index","preview_grid":"video-block"}',
            "slug": 1,
            "type": type
        }

        # define genre, if specified
        if genre: params['g'] = genre

        # return formatted url
        return "{}/{}?{}".format(SITE_URL, type, urllib.parse.urlencode(params))

    @staticmethod
    def process_movie_data(movie):
        """
        Processing of movie data before returning it in get_movie_data() function.

        Parameters:
            movie (dict): Full movie data from flixify. See flixify.get_movie_data().
        Returns:
            dict: Modified movie data.
        """
        try:

            # store the slug (last part of the url)
            movie["slug"] = movie["url"].replace("/movies/", "")

            # fix images (turn paths into full urls)
            BASE_URL = "https://a.flixify.com"
            images = movie["images"]
            previews = images["previews"]
            images["preview"] = BASE_URL + images["preview"]
            images["poster"] = BASE_URL + images["poster"]
            images["preview_large"] = BASE_URL + images["preview_large"]
            for img, path in enumerate(previews):
                previews[img] = BASE_URL + path

            if movie["certification"] is None: movie["certification"] = "N/A"
            if movie["year"] is None: movie["year"] = 0
            if movie["rating"] is None: movie["rating"] = 0

            if not "imdb_id" in movie: movie["imdb_id"] = ""


            return movie

        except: return False

    def get_movie_data(self, slug):
        """
        Gets data from flixify about a specific movie.

        Parameters:
            movie (dict): Item from items list returned by a download_data.

        Returns:
            dict: Full movie data from flixify. Example: https://pastebin.com/tW8idgyA
        """

        params = {
            "_t": self.var_t,
            "_u": self.var_u,
            "add_mroot": 1,
            "add_sequels": 1,
            "cast": 0,
            "crew": 0,
            "description": 1,
            "episodes_list": 1,
            "postersize": "poster",
            "previews": 1,
            "previewsizes": '{"preview_list":"big3-index","preview_grid":"video-block"}',
            "season_list": 1,
            "slug": 1,
            "sub": 1
        }

        # generate url
        url = SITE_URL + "movies/" + slug
        url += "?" + urllib.parse.urlencode(params)

        # send & parse request
        response = self.session.get(url, headers = self.headers)
        if response.status_code == 200:

            # parse movie data from response
            data = json.loads(response.text)
            movie = data["item"]

            # make any necessary modifications and return
            movie = flixify.process_movie_data(movie)
            return movie

        else: return False

    def download_data(self, type, page, genre = None):
        """
        Downloads json data from flixify api url, deserializes it into dict object.

        Parameters:
            type (string): Type of content. ('movies' or 'shows')
            page (int): Page to get items from.
            genre (string): Genre to get videos from.

        Returns:
            dict: A dictionary containing values from the response. Example data from download_data('movies', 1): https://pastebin.com/wq7PvT7F
        """

        url = self.build_list_url(type, page, genre)
        response = self.session.get(url, headers = self.headers)
        if response.status_code == 200:
            return json.loads(response.text)
        else: return False

def login(email, password):
    """Shortcut to flixify.flixify constructor from external modules."""
    return flixify(email, password)

class scraper(threading.Thread):

    def __init__(self, scraper):
        threading.Thread.__init__(self)
        self.scraper = scraper
        # note: this may need to automatically re-login eventually?

    def handle_movie(self, movie):
        """
        Uploads a movie to the database, if we don't already have it.

        Parameters:
            movie (dict): Movie data from list returned by flixify.download_data().
        """

        # skip movie if we already have it
        slug = movie["url"].replace("/movies/", "")
        print(slug)
        if database.get_movie(slug):
            return False

        # upload movie to database
        print("movie: " + slug)
        movie = self.scraper.get_movie_data(slug)
        if not movie:
            print("failed")
        else:
            database.upload_movie(movie)

    def run_once(self):
        """Loops through every single movie on flixify and stores it if we don't already have it."""

        for genre in GENRES:

            # start at page 1, keep incrementing until we break
            page = 1
            while True:

                print("genre: {}, page: {}\n".format(genre, page))
                data = self.scraper.download_data("movies", page, genre)

                # if we're out of videos
                if not data: break

                # make sure we have some movies
                movies = data['items']
                if len(movies) == 0: break

                for movie in movies:
                    self.handle_movie(movie)
                    print("------------------------------------\n")

                page += 1

    def run(self):
        """Runs run_once every hour to upload all flixify movie data."""

        # run scraper every hour
        while True:
            self.run_once()
            time.sleep(60 * 60)

class updater(threading.Thread):

    def __init__(self, scraper):
        threading.Thread.__init__(self)
        self.scraper = scraper

    def run_once(self):

        # loop through movies that need to be updated
        for movie in database.get_movies_pending_update():

            # get movie data from slug
            slug = movie[1]
            movie = self.scraper.get_movie_data(slug)

            # update the row in the database
            database.update_movie(slug, movie)

    def run(self):
        "Runs run_once every minute to update all movie links where update_required = 1 in database."

        # run update every minute
        while True:
            self.run_once()
            time.sleep(60)
